### PR TITLE
feat: support streaming reduce for Milvus-serverless(#31410)

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #31410

### CODE REVIEW

File: internal/querynodev2/services.go [hunk 1]
```diff
@@ -681,7 +681,13 @@ func (node *QueryNode) SearchSegments(ctx context.Context, req *querypb.SearchRe
 		return resp, nil
 	}
 
-	task := tasks.NewSearchTask(searchCtx, collection, node.manager, req, node.serverID)
+	var task tasks.Task
+	if paramtable.Get().QueryNodeCfg.UseStreamSearch.GetAsBool() {
+		task = tasks.NewStreamingSearchTask(searchCtx, collection, node.manager, req, node.serverID)
```
Comment (tedxu) [inline internal/querynodev2/services.go:686]: Is it necessary to introduce a new type of task here? The future maintenance work could be expensive.

File: internal/core/src/segcore/StreamReduce.h [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+    void
+    AssembleMergedResult();
+
+    std::vector<char>
+    GetSearchResultDataSlice(int slice_index);
+
+    void
+    CleanReduceStatus();
+
```
Comment (MrPresent-Han) [inline internal/core/src/segcore/StreamReduce.h:204]: Throughout the merging process, we keep a merged_search_result in the reducer and merging all following searchResults till the end

File: internal/querynodev2/segments/search.go [hunk 3]
```diff
@@ -109,6 +109,78 @@ func searchSegments(ctx context.Context, mgr *Manager, segments []Segment, segTy
...
+		resultStream <- searchResult
+		// update metrics
+		elapsed := tr.ElapseSpan().Milliseconds()
+		metrics.QueryNodeSQSegmentLatency.WithLabelValues(fmt.Sprint(paramtable.GetNodeID()),
+			metrics.SearchLabel, searchLabel).Observe(float64(elapsed))
+		metrics.QueryNodeSegmentSearchLatencyPerVector.WithLabelValues(fmt.Sprint(paramtable.GetNodeID()),
+			metrics.SearchLabel, searchLabel).Observe(float64(elapsed) / float64(searchReq.getNumOfQuery()))
+		return nil
+	}
+
```
Comment (MrPresent-Han) [inline internal/querynodev2/segments/search.go:157]: Serially at a certain concurrency in the search task and stream reduce these search results gradually.
The concurrency is defined by streamCredits which is a channel len=defaultStreamConcurrency, in the following pr, this concurrency will be aligned with the memory avaiable of LRU Cache

File: internal/querynodev2/services.go (hunk 1 above)
Comment (MrPresent-Han) [reply internal/querynodev2/services.go:686]: The benefits for separating SearchTasks and StreamingSearchTasks：
1. logically safe and clear, we can totally disaggrate logics of batch reduce and stream reduce
2. As there're NQ-merge logic in querynode, the search tasks that will be using stream reduced should be isolated from search-tasks to be batch reduced.

File: internal/querynodev2/tasks/task.go [hunk 4]
```diff
@@ -355,3 +357,263 @@ func (t *SearchTask) combinePlaceHolderGroups() {
...
+	searchReq, err := segments.NewSearchRequest(t.ctx, t.collection, req, t.placeholderGroup)
+	if err != nil {
+		return err
+	}
+	defer searchReq.Delete()
+
+	var pinnedSegments []segments.Segment
+	// 1. search&&reduce or streaming-search&&streaming-reduce
+	metricType := searchReq.Plan().GetMetricType()
+	if req.GetScope() == querypb.DataScope_Historical {
```
Comment (MrPresent-Han) [inline internal/querynodev2/tasks/task.go:465]: only stream reduce for sealed segments

**Review by MrPresent-Han — COMMENTED** [review]
need review

File: internal/core/src/segcore/StreamReduce.cpp [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+                     real_topKs.end(),
+                     search_result->topk_per_nq_prefix_sum_.begin() + 1);
+}
+
+void
+StreamReducerHelper::StreamReduceSearchResultForOneNQ(int64_t qi,
+                                                      int64_t topK,
+                                                      int64_t& offset) {
+    //1. clear heap for preceding left elements
+    while (!heap_.empty()) {
```
[the remark this answers is not in the document]
Comment (MrPresent-Han) [reply internal/core/src/segcore/StreamReduce.cpp:319]: True, I will change this in the following pr

File: internal/core/src/segcore/StreamReduce.h [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+ public:
+    void
+    MergeReduce();
+    void*
+    SerializeMergedResult();
+
+ protected:
+    void
+    FilterSearchResults();
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/StreamReduce.h:172]: We need to add a short comment for each fucntion
So it's easier for reviewers to understand

File: internal/core/src/segcore/StreamReduce.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaofan-luan) [reply internal/core/src/segcore/StreamReduce.cpp:319]: We don't really need such a large heap.
We only need to maintain a heap with K elements.

If heap.size() < k, we push result into heap. if heap size > k, we compare heap.top with candidates and decide which one should stay in the heap

[the remark this answers is not in the document]
Comment (xiaofan-luan) [reply internal/core/src/segcore/StreamReduce.cpp:319]: and we can keep this heap in merged_search_result so it doesn't need an extra heap and another pop out

File: internal/core/src/segcore/StreamReduce.cpp [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+#include "SegmentInterface.h"
+#include "segcore/Utils.h"
+#include "Reduce.h"
+#include "segcore/pkVisitor.h"
+#include "segcore/ReduceUtils.h"
+
+namespace milvus::segcore {
+
+void
+StreamReducerHelper::FillEntryData() {
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/StreamReduce.cpp:22]: Seems that we are trying to FillEntryData every time. which may takes too much time ios.

Ideally we should:

1. put segments into a group. (each group could roughly around 100GB disk usage)
2. search request by group and reduce it in one time
3. if search a lot of segment in one group, using streamReducer to help to reduce the disk usage

File: internal/querynodev2/segments/search.go [hunk 8]
```diff
@@ -109,6 +110,93 @@ func searchSegments(ctx context.Context, mgr *Manager, segments []Segment, segTy
...
+	errs := make([]error, len(segments))
+	printWithoutIndex := false
+	// calling segment search in goroutines
+	var wg sync.WaitGroup
+	for i, segment := range segments {
+		wg.Add(1)
+		go func(seg Segment, i int) {
```
[the remark this answers is not in the document]
Comment (MrPresent-Han) [reply internal/querynodev2/segments/search.go:161]: True， errgroup is  really more concise

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/core/src/segcore/StreamReduce.h (hunk 6 above)
Comment (MrPresent-Han) [reply internal/core/src/segcore/StreamReduce.h:174]: Well, for most methods in this head file, the name of method has been clear enough I suppose compared to one-sentence comment

**Review by MrPresent-Han — COMMENTED** [review]
before review for the second round

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Pull request closed without merging**